src/Interpreters/ProcessList.cpp

```diff
@@ -446,13 +455,15 @@ void QueryStatus::ExecutorHolder::remove()
     executor = nullptr;
 }
 
-CancellationCode QueryStatus::cancelQuery(bool)
+CancellationCode QueryStatus::cancelQuery(bool, CancelReason reason)
 {
     if (is_killed.load())
         return CancellationCode::CancelSent;
 
     is_killed.store(true);
 
+    cancel_reason = reason;
```

Comment:
here is a data race.
You need to use `atomic_exchange` here to set `cancel_reason` only at the first try.